modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -359,21 +440,105 @@ private long absFileIdx(CheckpointEntry pointer) {
         return tup;
     }
 
+    /**
+     * Search the earliest WAL pointer for particular group, matching by counter for partitions.
+     *
+     * @param grpId Group id.
+     * @param partsCounter Partition mapped to update counter.
+     * @return Earliest WAL pointer for group specified.
+     */
+    @Nullable public WALPointer searchEarliestWalPointer(int grpId, Map<Integer, Long> partsCounter) throws IgniteCheckedException {
+        if (F.isEmpty(partsCounter))
+            return null;
+
+        Map<Integer, Long> modifiedPartsCounter = new HashMap<>();
+
+        modifiedPartsCounter.putAll(partsCounter);
+
+        FileWALPointer minPtr = null;
+
+        for (Long cpTs : checkpoints(true)) {
+            CheckpointEntry cpEntry = entry(cpTs);
+
+            Iterator<Map.Entry<Integer, Long>> iter = modifiedPartsCounter.entrySet().iterator();
+
+            while (iter.hasNext()) {
+                Map.Entry<Integer, Long> entry = iter.next();
+
+                Long foundCntr = cpEntry.partitionCounter(cctx, grpId, entry.getKey());
+
+                if (foundCntr != null && foundCntr <= entry.getValue()) {
+                    iter.remove();
+
+                    FileWALPointer ptr = (FileWALPointer)cpEntry.checkpointMark();
+
+                    if (ptr == null) {
+                        throw new IgniteCheckedException("Could not find start pointer for partition [part="
+                            + entry.getKey() + ", partCntrSince=" + entry.getValue() + "]");
+                    }
+
+                    if (minPtr == null || ptr.compareTo(minPtr) < 0)
+                        minPtr = ptr;
+                }
+
+            }
```

Comment:
Can we break; here if modifiedPartsCounter is already empty?

Reply:
Done.